wicketd/src/artifacts/update_plan.rs

```diff
@@ -0,0 +1,1016 @@
+    pub(super) fn new(
+        system_version: SemverVersion,
+        log: &'a Logger,
+    ) -> Result<Self, RepositoryError> {
+        let extracted_artifacts = ExtractedArtifacts::new(log)?;
+        Ok(Self {
+            system_version,
+            gimlet_sp: BTreeMap::new(),
+            gimlet_rot_a: None,
+            gimlet_rot_b: None,
+            psc_sp: BTreeMap::new(),
+            psc_rot_a: None,
+            psc_rot_b: None,
+            sidecar_sp: BTreeMap::new(),
+            sidecar_rot_a: None,
+            sidecar_rot_b: None,
+            host_phase_1: None,
+            trampoline_phase_1: None,
+            trampoline_phase_2: None,
+            host_phase_2_hash: None,
+            control_plane_hash: None,
+
+            extracted_artifacts,
+            log,
+        })
+    }
+
+    pub(super) fn add_artifact(
+        &mut self,
+        artifact_id: ArtifactId,
+        artifact_hash: ArtifactHash,
+        reader: io::BufReader<impl io::Read>,
+        by_id: &mut BTreeMap<ArtifactId, Vec<ArtifactHashId>>,
+        by_hash: &mut HashMap<ArtifactHashId, ExtractedArtifactDataHandle>,
+    ) -> Result<(), RepositoryError> {
+        // If we don't know this artifact kind, we'll still serve it up by hash,
+        // but we don't do any further processing on it.
+        let Some(artifact_kind) = artifact_id.kind.to_known() else {
+            return self.add_unknown_artifact(
+                artifact_id,
+                artifact_hash,
+                reader,
+                by_id,
+                by_hash,
+            );
+        };
+
+        match artifact_kind {
+            KnownArtifactKind::GimletSp
+            | KnownArtifactKind::PscSp
+            | KnownArtifactKind::SwitchSp => self.add_sp_artifact(
+                artifact_id,
+                artifact_kind,
+                artifact_hash,
+                reader,
+                by_id,
+                by_hash,
+            ),
+            KnownArtifactKind::GimletRot
+            | KnownArtifactKind::PscRot
+            | KnownArtifactKind::SwitchRot => self.add_rot_artifact(
+                artifact_id,
+                artifact_kind,
+                reader,
+                by_id,
+                by_hash,
+            ),
+            KnownArtifactKind::Host => self.add_host_artifact(
+                artifact_id,
+                artifact_kind,
+                reader,
+                by_id,
+                by_hash,
+            ),
+            KnownArtifactKind::Trampoline => self.add_trampoline_artifact(
+                artifact_id,
+                artifact_kind,
+                reader,
+                by_id,
+                by_hash,
+            ),
+            KnownArtifactKind::ControlPlane => self.add_control_plane_artifact(
+                artifact_id,
+                artifact_kind,
+                artifact_hash,
+                reader,
+                by_id,
+                by_hash,
+            ),
```

Comment:
Why pass in the `KnownArtifactKind` for these if we're always going to ensure it's one particular kind anyway?

Reply:
We do use it in the SP / RoT methods (since we need to known which of the gimlet/sidecar/psc kinds it is), and I just copied that over. Removed the arg from the host/trampoline/control plane methods in 88393ed9